### DISCUSSION TITLE
Allow `since` as argument of put pattern

### DISCUSSION THREAD

**Discussion by waalge** [CardanoSolutions/kupo#27] [Archives]
Here's my narration of how I think kupo works. Please correct me where its apparent I don't know what I'm talking about. 

On start, kupo looks at all blocks since the `--since` arg ( defaulting to origin) for matches on `--match`  args.
Kupo reads blocks sequentially (caveat: rollbacks) in a single sequence. 
On restarting, kupo will go the last checkpoint (?) and continue syncing. 
Kupo allows adding and deleting patterns while running. 
However, matches will only be indexed for a given pattern in a given block if  the  pattern is added before the  block has been read by kupo. 

If there is some delay in learning of a new "interesting" address, and kupo has already read this block there is no way to add this address and get a complete utxo set for it.  
 
How bad would it be to have kupo be able to handle multiple block readings?

**Comment by KtorZ** [comment]
Okay. So, it would be feasible to instrument the internal chain consumer to also synchronize from a given point when changing patterns.

What isn't possible (at least in the current design à without a major overhaul) is to have synchronization going on at multiple levels on-chain. Said differently, in order to synchronize something from the past, Kupo will have to rollback the entire database from that point on. The synchronization isn't per pattern, there's only one synchronisation thread.

**Reply by waalge** [reply]
>  at least in the current design without a major overhaul

Hmmmm. That is what i thought might be the case. 

Ogmios and a Node would be fine with multiple connections, right? Just kupo.

I'd be interested to know if anyone else thinks this would be useful.

Cheers!

**Reply by KtorZ** [reply]
Ogmios and node would be fine handling multiple connections, yes... BUT. With the default settings, they actually both limit their resources to 2 cores only. And, if you observe the node during a full chain-sync, you'll see that the CPU usage approaches 200%. Thus, multiplying the clients (e.g., starting multiple instances of Kupo) will work but the node will have to divide the load between the two processes. Works fine, just a bit slower. 

For the record, I've seen a single local node serve 100+ client connections. It's slow, but it works, eventually.

**Reply by bakon11** [reply]
Not sure if I'm understanding the idea of `since` here. 

But for example, I'm using kupo to build a full node wallet and of course I'd only need to sync addresses/patterns added to the wallet.

So could `since` be used here, when adding a new pattern that has existing UTXOs you could tell it to sync and append all historical data of that new pattern to existing kupo DB?

My work around idea was when an address/pattern has been added at a later date, that has a utxo history from previous blocks that kupo already synced. Is to start a new instance of kupo, sync that address/pattern to a new DB and then just merge the two DB's together.

**Reply by waalge** [reply]
> Not sure if I'm understanding the idea of since here.

I have some bound on how long ago a new interesting address has been in use for. And so I'd like kupo to re-read blocks since that point, And preferably while maintaining the up-to-date-ness of matches on existing patterns.

>  and then just merge the two DB's

Perhaps this is that simple. I'm not sure i'm brave enough to try that.

@KtorZ Thanks for those insights. 

> It's slow, but it works, eventually.

I think as long as kupo can tell you how up-to-date the sync of a given pattern is, this is OK. 

At present and for the immediate future this feature is not a must-have: In a v0 use case everything is centralized and my machine handles the construction of most things including producing the validators, and so too the new interesting address, for a user. Thus I should know the address before its used. 
If/when this is handed off to the clients machine, then this would potentially be a painpoint/ hard wall. I think. 

I just wanted to put this out there to see a) if there was a easy, neat mod to do this that I'd overlooked b) see if there's any other interested.
And I'd be happy to work on it, but I don't want to spoil your far-superior-to-mine haskell code.

**Reply by KtorZ** [reply]
One thing to keep in mind too, is that Kupo process between 1000-10000 blocks per second (depending on how full they are). Which means that rewinding the index to some blocks in the past, might it be an entire epoch is still relatively fast. Thus, the (complexity) cost of independently side-syncing a certain address compared to just rolling back the index is perhaps not worth it.

**Reply by bakon11** [reply]
> One thing to keep in mind too, is that Kupo process between 1000-10000 blocks per second (depending on how full they are). Which means that rewinding the index to some blocks in the past, might it be an entire epoch is still relatively fast. Thus, the (complexity) cost of independently side-syncing a certain address compared to just rolling back the index is perhaps not worth it.

True, if it can sync the whole ledger with all addresses in `55min` I think from thread #20, only a handful would be a synch hehe.

**Reply by waalge** [reply]
> perhaps not worth it

That seems reasonable

In either case I guess you still run back into the problem of not being able to specify `--since` by time alone.
You also need the block hash, right?

**Reply by KtorZ** [reply]
You need the block hash. But the new API coming in v2.0.0 gives you ways to retrieve any block hash associated with slot numbers. So in principle, you can easily recover any block hash.

**Reply by waalge** [reply]
Nice. Is the spec or roadmap for v2 public?

**Reply by KtorZ** [reply]
The user manual, spec and changelog are updated along the way. The roadmap itself is maintained as a github project, visible here: https://github.com/CardanoSolutions/kupo/projects/2

**Reply by waalge** [reply]
Its currently content-less for me

**Reply by KtorZ** [reply]
Funny. I think it's a Github bug here. I also have nothing if I click on the link directly. Try to click on the project from: 

https://github.com/CardanoSolutions/kupo/projects

:shrug:

**Reply by waalge** [reply]
:) That one works. I'm not familiar with the github feature set
